auth/server/base/src/main/java/org/wildfly/security/authz/RegexRoles.java

```diff
@@ -0,0 +1,116 @@
+/*
+ * JBoss, Home of Professional Open Source.
+ * Copyright 2019 Red Hat, Inc., and individual contributors
+ * as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.wildfly.security.authz;
+
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+import java.util.regex.Pattern;
+import java.util.regex.PatternSyntaxException;
+import java.util.stream.StreamSupport;
+
+import static org.wildfly.security.auth.server._private.ElytronMessages.log;
+
+/**
+ * A regex roles.
+ * <p>
+ * Roles that are checked against pattern. Roles that match the pattern are then mapped with given replacement. Pattern can capture groups that replacement can make use of.
+ *
+ * @author <a href="mailto:dvilkola@redhat.com">Diana Vilkolakova</a>
+ */
+class RegexRoles implements Roles {
+    private final Roles delegate;
+    private final Pattern pattern;
+    private final String replace;
+    private final Boolean keepNonMapped;
+    private final Boolean replaceAll;
+
+    RegexRoles(final Roles delegate, Pattern pattern, final String replace, final Boolean keepNonMapped, final boolean replaceAll) {
+        this.delegate = delegate;
+        this.pattern = pattern;
+        this.replace = replace;
+        this.keepNonMapped = keepNonMapped;
+        this.replaceAll = replaceAll;
+    }
+
+    public boolean contains(final String roleName) {
+        try {
+            return StreamSupport.stream(delegate.spliterator(), false).anyMatch(role ->
+            {
+                String pattern = this.pattern.pattern();
+                Boolean matches = this.pattern.matcher(role).matches();
+
+                boolean doesNotMatchRegexButRoleExists = !matches && keepNonMapped && role.equals(roleName);
+                boolean matchesRegexAndRoleExists = false;
+                if (matches && replaceAll) {
```

Comment:
I think this `if` block can be simplified using ```this.pattern.matcher(role).replaceAll(replace)``` and ```this.pattern.matcher(role).replaceFirst(replace)```. I don't think the while loop is needed.

Reply:
@fjuma I added while loop for situation in [this test case](https://github.com/Skyllarr/wildfly-elytron/blob/ELY-regex-role-mapper/auth/server/base/src/test/java/org/wildfly/security/authz/RegexRoleMapperTest.java#L174). 

Here I want to replace all substrings `guest` with substring `user`. Problem is that result of `this.pattern.matcher("app-guest-first-time-guest").replaceAll("$1user$2")`  is `app-guest-first-time-user` and not `app-user-first-time-user`. This is because `this.pattern` is not `guest` but `(.*)guest(.*)` in order to match the role in the first place. 

But this is maybe not intuitive behaviour and `replaceAll` does not make sense here?

Reply:
Ah, I see now. For this test, I think `pattern` should be "guest" and `replace` should be "user". Instead of using `matches()`, I think you should be able to use `find()` instead since that returns `true` iff a subsequence of the input sequence matches this matcher's pattern.

Reply:
@fjuma Thank you for help. I updated all PRs and did use `find()` instead.